packages/compile-common/src/compilerSupplier/strategy.ts

```diff
@@ -0,0 +1,106 @@
+import type { Constructor as FreeConstructor } from "./constructor";
+import type { Results as FreeResults } from "./results";
+
+export namespace Strategy {
+  /**
+   * Type-level description of a compiler supplier strategy
+   */
+  export type Specification = {
+    constructor: Omit<FreeConstructor.Specification, "result">;
+    results: FreeResults.Specification;
+    allowsLoadingSpecificVersion: boolean;
+    allowsListingVersions: boolean;
+  };
+
+  /**
+   * Scoped re-exports of Constructor, using Strategy.Specification as
+   * generic param
+   */
+  export namespace Constructor {
+    /**
+     * Getter for the specified constructor, intersected with fixed result
+     * type of Strategy<S>
+     */
+    export type Specification<S extends Strategy.Specification> =
+      S["constructor"] & { result: Strategy<S>; };
+
+    /**
+     * Constructor options argument for a specified strategy
+     */
+    export type Options<S extends Strategy.Specification> = FreeConstructor.Options<
+      Strategy.Constructor.Specification<S>
+    >;
+  }
+
+  /**
+   * Type representing a constructor for specified strategy
+   */
+  export type Constructor<S extends Strategy.Specification> = FreeConstructor<
+    Strategy.Constructor.Specification<S>
+  >;
+
+  /**
+   * Results types for a particular strategy
+   */
+  export namespace Results {
+    export type Specification<S extends Strategy.Specification> =
+      S["results"];
+
+    export type Load<S extends Strategy.Specification> = FreeResults.Load<
+      Strategy.Results.Specification<S>
+    >;
+
+    export type List<S extends Strategy.Specification> = FreeResults.List<
+      Strategy.Results.Specification<S>
+    >;
+  }
+
+  /**
+   * Getter type for whether specified strategy allows argument to load()
+   */
+  export type AllowsLoadingSpecificVersion<
+    S extends Strategy.Specification
+  > = S["allowsLoadingSpecificVersion"];
+
+  /**
+   * Getter type for whether specified strategy provides version listing
+   */
+  export type AllowsListingVersions<
+    S extends Strategy.Specification
+  > = S["allowsListingVersions"];
+
+}
+
+/**
+ * An object of type Strategy<S> provides version loading and possibly version
+ * listing functionality
+ */
+export type Strategy<S extends Strategy.Specification> = BaseStrategy<S> &
+  (true extends Strategy.AllowsLoadingSpecificVersion<S>
+    ? {
+        load(version?: string): Strategy.Results.Load<S>;
+      }
+    : {
+        load(): Strategy.Results.Load<S>;
+      }) &
+  (true extends Strategy.AllowsListingVersions<S>
+    ? {
+        list(): Strategy.Results.List<S>;
+      }
+    : {});
```

Comment:
There's some copy-and-paste here... should these component types get individual names to reduce that?

Reply:
I'm not sure what copy-and-paste you're talking about?

Reply:
Like, comparing the definition of `Strategy` and the definition of `BaseStrategy`, where in both cases you give these conditions for them to also have the types
`{ load(version?: string): Strategy.Results.Load<S>; }` and `{ list(): Strategy.Results.List<S>; }`.  Should those types have names?

Reply:
Hm, yeah ok! How do you like this: https://github.com/trufflesuite/truffle/pull/4224/commits/bdd85cc7271f9af290f121ad0b32bb8bec753cdc